# HG changeset patch
# User Jacob Dawid <jacob.dawid@googlemail.com>
# Date 1302762404 -7200
#      Thu Apr 14 08:26:44 2011 +0200
# Node ID 73af3d75ddcd6023dfa28446192a8c9c276393b8
# Parent  785516f65f0ae1ca8a47ddad965a68feb226033b
Variables now get assorted by scope.

diff --git a/gui//src/VariablesDockWidget.cpp b/gui//src/VariablesDockWidget.cpp
--- a/gui//src/VariablesDockWidget.cpp
+++ b/gui//src/VariablesDockWidget.cpp
@@ -52,19 +52,45 @@ void VariablesDockWidget::updateTreeEntr
         Matrix matrixValue = symbolRecord.varval().matrix_value();
         // TODO: Display matrix.
     } else {
         treeItem->setData(2, 0, QString("<Type not recognized>"));
     }
 }
 
 void VariablesDockWidget::setVariablesList(QList<SymbolRecord> symbolTable) {
+
+    // Split the symbol table into its different scopes.
+    QList<SymbolRecord> localSymbolTable;
+    QList<SymbolRecord> globalSymbolTable;
+    QList<SymbolRecord> persistentSymbolTable;
+
+    foreach(SymbolRecord symbolRecord, symbolTable) {
+        if(symbolRecord.is_local()) {
+            localSymbolTable.append(symbolRecord);
+        }
+
+        if(symbolRecord.is_global()) {
+            globalSymbolTable.append(symbolRecord);
+        }
+
+        if(symbolRecord.is_persistent()) {
+            persistentSymbolTable.append(symbolRecord);
+        }
+    }
+
+    updateScope(0, localSymbolTable);
+    updateScope(1, globalSymbolTable);
+    updateScope(2, persistentSymbolTable);
+}
+
+void VariablesDockWidget::updateScope(int topLevelItemIndex, QList<SymbolRecord> symbolTable) {
     // This method may be a little bit confusing; variablesList is a complete list of all
     // variables that are in the workspace currently.
-    QTreeWidgetItem *topLevelItem = m_variablesTreeWidget->topLevelItem(0);
+    QTreeWidgetItem *topLevelItem = m_variablesTreeWidget->topLevelItem(topLevelItemIndex);
 
     // First we check, if any variables that exist in the model tree have to be updated
     // or created. So we walk the variablesList check against the tree.
     foreach(SymbolRecord symbolRecord, symbolTable) {
         int childCount = topLevelItem->childCount();
         bool alreadyExists = false;
         QTreeWidgetItem *child;
 
diff --git a/gui//src/VariablesDockWidget.h b/gui//src/VariablesDockWidget.h
--- a/gui//src/VariablesDockWidget.h
+++ b/gui//src/VariablesDockWidget.h
@@ -4,18 +4,18 @@
 #include <QDockWidget>
 #include <QTreeWidget>
 #include "OctaveLink.h"
 
 class VariablesDockWidget : public QDockWidget
 {
 public:
     VariablesDockWidget(QWidget *parent = 0);
-    void setVariablesList(QList<SymbolRecord> variablesList);
+    void setVariablesList(QList<SymbolRecord> symbolTable);
 
 private:
     void construct();
     void updateTreeEntry(QTreeWidgetItem *treeItem, SymbolRecord symbolRecord);
-
+    void updateScope(int topLevelItemIndex, QList<SymbolRecord> symbolTable);
     QTreeWidget *m_variablesTreeWidget;
 };
 
 #endif // VARIABLESDOCKWIDGET_H
